File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.cc
#include <iostream>
#include <vector>
#include <string>
#include <queue>
using namespace std;

typedef struct node {
	int i;
	int j;
	int length;
	bool brokenBrick;
}Node;

typedef struct brick {
	int i;
	int j;
}Brick;


vector<vector<vector<bool>>> visited;
vector<Brick> bricks;
vector<vector<int>> map;

bool isArrived = false;

int dir[][2] = { {0,1},{0,-1},{1,0},{-1,0} };

bool isInBoundary(int i, int j) {
	return 0 <= i && i < map.size() && 0 <= j && j < map[0].size();
}

int minLength = 2100000000;

void bfs() {



	queue<Node> nextNode;
	nextNode.push({ 0,0,1,false });


	visited[0][0][0] = true;

	while (nextNode.empty() != true) {

		Node curNode = nextNode.front(); nextNode.pop();

		if (curNode.i == map.size() - 1 && curNode.j == map[0].size() - 1) {
			if (minLength > curNode.length)
				minLength = curNode.length;
			isArrived = true;
		}
		else {
			for (int i = 0; i < 4; i++) {
				int nextI = curNode.i + dir[i][0];
				int nextJ = curNode.j + dir[i][1];
				if (isInBoundary(nextI, nextJ)) {

					if (curNode.brokenBrick == true) {

						if (map[nextI][nextJ] == 0 && visited[nextI][nextJ][1] == false) {
							visited[nextI][nextJ][1] = true;
							nextNode.push({ nextI, nextJ, curNode.length + 1,curNode.brokenBrick });
						}

					}
					else {
						if (map[nextI][nextJ] == 0) {
							if (visited[nextI][nextJ][0] == false) {
								visited[nextI][nextJ][0] = true;
								nextNode.push({ nextI, nextJ, curNode.length + 1, curNode.brokenBrick });
							}
						}
						else {
							if (visited[nextI][nextJ][1] == false) {
								visited[nextI][nextJ][1] = true;
								nextNode.push({ nextI, nextJ, curNode.length + 1, true });
							}
						}
					}
				}

			}
		}
	}
}

int main() {

	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n, m; cin >> n >> m;

	for (int i = 0; i < n; i++) {
		vector<int> temp;
		vector<vector<bool>> temp2;
		string input; cin >> input;
		for (int j = 0; j < m; j++) {
			temp.push_back(input[j] - '0');
			if (input[j] - '0' == 1)
				bricks.push_back({ i,j });
			vector<bool> temp3 = { false,false };
			temp2.push_back(temp3);
		}
		map.push_back(temp);
		visited.push_back(temp2);
	}

	
	bfs();
	

	if (isArrived == true)
		cout << minLength;
	else
		cout << -1;
	
}

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    private static BufferedReader br;
    private static StringTokenizer st;


    private static int N,M;
    private static int[][] map;

    static class Node{
        int i;
        int j;
        int length;
        boolean broken;

        public Node(int i, int j, int length, boolean broken){
            this.i = i;
            this.j = j;
            this.length = length;
            this.broken = broken;
        }
    }

    private static boolean isInBoundary(int i, int j){
        return 0 <= i && i < N && 0 <= j && j < M;
    }


    private static int getAnswer(){

        boolean[][][] visited = new boolean[N][M][2];
        Queue<Node> nextNode = new LinkedList<>();
        int minLength = Integer.MAX_VALUE;
        boolean isArrived = false;
        int[][] dir = {{1,0},{-1,0},{0,1},{0,-1}};

        nextNode.add(new Node(0,0,1,false));
        visited[0][0][0] = true;

        while(nextNode.isEmpty() != true){

            Node curNode = nextNode.poll();

            if(curNode.i == N-1 && curNode.j == M-1){
                isArrived = true;
                minLength = Math.min(minLength, curNode.length);
            }
            else {
                for (int i = 0; i < 4; i++) {
                    int nextI = curNode.i + dir[i][0];
                    int nextJ = curNode.j + dir[i][1];
                    int nextLength = curNode.length + 1;
                    boolean nextBroken = curNode.broken;

                    if (isInBoundary(nextI, nextJ) == false)
                        continue;
                    if(nextLength > minLength)
                        continue;;

                        /*
                            1. 벽을 깨본적이 있고, 현재위치가 벽이 아닐때.
                            2. 벽을 깨본적이 없고,
                                2-1 현재위치가 벽일때
                                2-2 현재위치가 벽이 아닐때.
                         */
                    if(nextBroken == true){

                        if(map[nextI][nextJ] == 0){

                            if(visited[nextI][nextJ][1] == true)
                                continue;

                            nextNode.add(new Node(nextI,nextJ,nextLength,true));
                            visited[nextI][nextJ][1] = true;
                        }

                    }
                    else{

                        if(map[nextI][nextJ] == 0){

                            if(visited[nextI][nextJ][0] == true)
                                continue;

                            nextNode.add(new Node(nextI,nextJ,nextLength,false));
                            visited[nextI][nextJ][0] = true;
                        }

                        if(map[nextI][nextJ] == 1){
                            if(visited[nextI][nextJ][1] == true)
                                continue;

                            nextNode.add(new Node(nextI,nextJ,nextLength,true));
                            visited[nextI][nextJ][1] = true;
                        }
                    }

                }
            }
        }
        if(isArrived == true){
            return minLength;
        }
        else
            return -1;
    }

    public static void main(String args[]) throws IOException {

        br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        map = new int[N][M];

        for(int i = 0; i < N; i++) {
            String input = br.readLine();
            for (int j = 0; j < M; j++) {
                map[i][j] = input.charAt(j) - '0';
            }
        }

        System.out.println(getAnswer());
    }

}
